sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -153,6 +153,15 @@ object SetOperationPushDown extends Rule[LogicalPlan] with PredicateHelper {
         )
       )
 
+    // Push down limit into union
+    case Limit(exp, Union(left, right)) =>
+      Limit(exp,
+        Union(
+          CombineLimits(Limit(exp, left)),
+          CombineLimits(Limit(exp, right))
+        )
+      )
```

Comment:
You are right. `Limit` might not converge to the same position after multiple pushdown.  

Let me think about it. Thank you!